pkg/dockerregistry/server/manifestservice.go

```diff
@@ -0,0 +1,235 @@
+		return nil, err
+	}
+
+	if len(image.DockerImageManifest) == 0 {
+		// We don't have the manifest in the storage and we don't have the manifest
+		// inside the image so there is no point to continue.
+		return nil, distribution.ErrManifestUnknownRevision{
+			Name:     m.repo.Named().Name(),
+			Revision: dgst,
+		}
+	}
+
+	manifest, err = m.repo.manifestFromImageWithCachedLayers(image, ref.Exact())
+
+	if _, err := m.manifests.Put(WithRepository(ctx, m.repo), manifest, options...); err != nil {
+		context.GetLogger(m.ctx).Errorf("unable to put manifest to storage: %v", err)
+	} else if len(image.DockerImageManifest) > 0 {
+		image.DockerImageManifest = ""
+		image.DockerImageConfig = ""
+
+		if _, err := m.repo.updateImage(image); err != nil {
+			context.GetLogger(m.ctx).Errorf("error updating Image: %v", err)
+		}
+	}
+
+	return manifest, err
+}
+
+// Put creates or updates the named manifest.
+func (m *manifestService) Put(ctx context.Context, manifest distribution.Manifest, options ...distribution.ManifestServiceOption) (digest.Digest, error) {
+	context.GetLogger(m.ctx).Debugf("(*manifestService).Put")
+
+	mh, err := NewManifestHandler(m.repo, manifest)
+	if err != nil {
+		return "", regapi.ErrorCodeManifestInvalid.WithDetail(err)
+	}
+	mediaType, payload, canonical, err := mh.Payload()
+	if err != nil {
+		return "", regapi.ErrorCodeManifestInvalid.WithDetail(err)
+	}
+
+	if !m.acceptschema2 && mediaType == schema2.MediaTypeManifest {
+		err = fmt.Errorf("manifest V2 schema 2 not allowed")
+		return "", regapi.ErrorCodeManifestInvalid.WithDetail(err)
+	}
+
+	_, err = m.manifests.Put(WithRepository(ctx, m.repo), manifest, options...)
+	if err != nil {
+		return "", err
+	}
+
+	// Calculate digest
+	dgst := digest.FromBytes(canonical)
+
+	// Upload to openshift
+	ism := imageapi.ImageStreamMapping{
+		ObjectMeta: kapi.ObjectMeta{
+			Namespace: m.repo.namespace,
+			Name:      m.repo.name,
+		},
+		Image: imageapi.Image{
+			ObjectMeta: kapi.ObjectMeta{
+				Name: dgst.String(),
+				Annotations: map[string]string{
+					imageapi.ManagedByOpenShiftAnnotation: "true",
+				},
+			},
+			DockerImageReference:         fmt.Sprintf("%s/%s/%s@%s", m.repo.registryAddr, m.repo.namespace, m.repo.name, dgst.String()),
+			DockerImageManifest:          string(payload),
+			DockerImageManifestMediaType: mediaType,
+		},
+	}
+
+	for _, option := range options {
+		if opt, ok := option.(distribution.WithTagOption); ok {
+			ism.Tag = opt.Tag
+			break
+		}
+	}
+
+	if err = mh.FillImageMetadata(ctx, &ism.Image); err != nil {
+		return "", err
+	}
+
+	// Remove the raw manifest as it's very big and this leads to a large memory consumption in etcd.
+	ism.Image.DockerImageManifest = ""
+	ism.Image.DockerImageConfig = ""
+
+	if err = m.repo.registryOSClient.ImageStreamMappings(m.repo.namespace).Create(&ism); err != nil {
+		// if the error was that the image stream wasn't found, try to auto provision it
```

Comment:
Are we sure this error is only because of the lack of an image stream or exceeded quota?